pytorch_lightning/trainer/connectors/logger_connector/result.py

```diff
@@ -564,7 +568,9 @@ def any_tensor(_: Any) -> None:
                 metrics["log"][forked_name] = value
 
             # populate callback metrics. callback metrics don't take `_step` forked metrics
-            if self.training or result_metric.meta.on_epoch and not on_step:
+            if (self.training or result_metric.meta.on_epoch and not on_step) and (
+                dataloader_idx is None or result_metric.meta.dataloader_idx == dataloader_idx
+            ):
```

Comment:
I don't think this should be here, instead it should be

```py
    def valid_items(self, dataloader_idx: Optional[int] = None) -> Generator:
        """This function is used to iterate over current valid metrics."""
        return (
            (k, v)
            for k, v in self.items()
            if not (isinstance(v, ResultMetric) and v.has_reset)
            and (dataloader_idx in (None, v.meta.dataloader_idx))
        )
```

where this function calls

```py
for _, result_metric in self.valid_items(dataloader_idx=dataloader_idx):
```

Reply:
will update!